README
HTML-GenerateUtil version 1.11
==============================

Provides a number of functions that make generating HTML output
easier and faster. All written in XS for speed.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

COPYRIGHT AND LICENCE

Copyright (C) 2004-2011 by Opera Software Australia Pty Ltd

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 



Changes
Revision history for Perl extension HTML::GenerateUtil.

1.11  Fri Jul 15 17:55:00 2011
  - Document $H object with AUTOLOAD functions

1.10  Mon May 22 14:51:00 2009
  - Allow single argument version of escape_html()

1.09  Thu Mar 13 16:00:00 2008
  - Add an AUTOLOAD method which calls generate_tag

1.08  Fri Oct 19 17:09:00 2007
	- if you use s/(...)/escape_uri($1)/, then $1 might not be SvOK()
    because it's magic. Need to call SvGETMAGIC first

1.07  Fri Oct 28 15:34:00 2005
	- add escape_uri function
  - fix random but very rare cluck bug due to uninitialised memory usage

1.06  Thu Nov 11 16:17:00 2004
	- allow passing of array refs and hash refs as values to generate_attributes

1.05  Thu Jun 23 14:36:00 2004
	- fix crash when passing non-strings in values to generate_tag

1.04  Thu Jun 15 09:42:00 2004
	- fix memory overrun when EH_LEAVEKNOWN and EH_INPLACE used together

1.03  Thu Jun 14 17:12:00 2004
	- add "leaveknown" entities option
  - add "closetag" option

1.02  Thu May 27 11:42:00 2004
	- remove dependence on 5.8.4
  - move :consts import to first

1.01  Thu May 27 11:20:00 2004
	- allow array ref for unquoted values in generate_attribtes()
  - turn on optimisation again

1.00  Wed May 26 12:23:00 2004
	- initial release



GenerateFunctions.c

#include "EXTERN.h"
#include "perl.h"
#include "XSUB.h"

#include "ppport.h"
#include "GenerateFunctions.h"

SV * GF_escape_html(SV * str, int b_inplace, int b_lftobr, int b_sptonbsp, int b_leaveknown) {
  int i, maxentitylen = 0;
  STRLEN origlen, extrachars;
  char * sp, *newsp, c, lastc;
  SV * newstr;

  /* Get string pointer and length (in bytes) */
  if (b_inplace) {
    sp = SvPV_force(str, origlen);
  } else {
    sp = SvPV(str, origlen);
  }

  /* Calculate extra space required */
  extrachars = 0;
  c = '\0';
  for (i = 0; i < origlen; i++) {
    /* Need to keep track of previous char for '  ' => ' &nbsp;' expansion */
    lastc = c;
    c = sp[i];
    if (c == '<' || c == '>')
	    extrachars += 3;
    else if (c == '&' && (!b_leaveknown || !GF_is_known_entity(sp, i, origlen, &maxentitylen)))
	    extrachars += 4;
    else if (c == '"')
	    extrachars += 5;
    else if (b_lftobr && c == '\n')
	    extrachars += 3;
    else if (b_sptonbsp && c == ' ' && lastc == ' ') {
      extrachars += 5;
      /* don't pick up immediately again */
      c = '\0';
    }
  }

  /* Special single space case */
  if (b_sptonbsp && origlen == 1 && sp[0] == ' ') {
    extrachars += 5;
  }

  /*
   * Include maxentitylen in extrachars. Since in the actual substitution
   * phase, we work backwards copying characters towards the end of the
   * string as we go, we might overwrite part of an entity, and then try
   * and call GF_is_known_entity() on the string, which searches forward,
   * and then fails because we already overwrote the entity. So we always
   * make sure we've got maxentitylen extra chars, and then use the perl
   * OOK hack to offset the start of the string at the end
   */
  if (b_inplace) extrachars += maxentitylen;

  /* Create new SV, or grow existing SV */
  if (b_inplace) {
    newstr = str;
    SvGROW(newstr, origlen + extrachars + 1);
  } else {
    newstr = newSV(origlen + extrachars + 1);
    SvPOK_on(newstr);
    /* Make new string UTF-8 if input string was UTF-8 */
    if (SvUTF8(str))
      SvUTF8_on(newstr);
  }

  /* Set the length of the string */
  SvCUR_set(newstr, origlen + extrachars);

  /* Now do actual replacement (need to work
     backward for inplace change to work */

  /* Original string might have moved due to grow */
  sp = SvPV_nolen(str);

  /* Null terminate new string */
  newsp = SvPV_nolen(newstr) + origlen + extrachars;
  *newsp = '\0';

  c = '\0';
  for (i = origlen-1; i >= 0; i--) {
    lastc = c;
    c = sp[i];
    if (c == '<') {
      newsp -= 4;
      memcpy(newsp, "&lt;", 4);
    }
    else if (c == '>') {
      newsp -= 4;
      memcpy(newsp, "&gt;", 4);
    }
    else if (c == '&' && (!b_leaveknown || !GF_is_known_entity(sp, i, origlen, 0))) {
      newsp -= 5;
      memcpy(newsp, "&amp;", 5);
    }
    else if (c == '"') {
      newsp -= 6;
      memcpy(newsp, "&quot;", 6);
    }
    else if (b_lftobr && c == '\n') {
      newsp -= 4;
      memcpy(newsp, "<br>", 4);
    }
    else if (b_sptonbsp && c == ' ' && lastc == ' ') {
      newsp -= 6;
      memcpy(newsp, "&nbsp; ", 7);
      /* don't pick up immediately again */
      c = '\0';
    }
    else
      *--newsp = c;
  }

  /* Special single space case */
  if (b_sptonbsp && origlen == 1 && sp[0] == ' ') {
    newsp -= 5;
    memcpy(newsp, "&nbsp;", 6);
  }

  if (b_inplace && maxentitylen)
    sv_chop(newstr, newsp);

  if (SvPV_nolen(newstr) != newsp) {
    croak("Unexpected length mismatch");
    return 0;
  }

  return newstr;
}

SV * GF_generate_attributes(HV * attrhv) {
  int i, j, estimatedlen = 1;
  I32 keylen;
  char * key, tmp[64];
  SV * attrstr, * val;

  /* Iterate through keys to work out an estimated final length */
  while ((val = hv_iternextsv(attrhv, &key, &keylen))) {
    estimatedlen += keylen + 1;
    estimatedlen += GF_estimate_attribute_value_len(val) + 3;
  }

  /* warn("estimated len: %d", estimatedlen); */

  attrstr = newSV(estimatedlen);
  SvPOK_on(attrstr);

  /* Now iteratre and build actual string */
  hv_iterinit(attrhv);
  while ((val = hv_iternextsv(attrhv, &key, &keylen))) {

    /* Add space to string if already something in it */
    if (SvCUR(attrstr))
      sv_catpvn(attrstr, " ", 1);

    /* For key, convert to lower case and add to attrstr */
    if (keylen < 64) {
      /* If key starts with - (eg -width => '10%'), skip - */
      j = 0;
      i = (keylen && key[0] == '-' ? 1 : 0);
      for (; i < keylen; i++)
        tmp[j++] = toLOWER(key[i]);
      sv_catpvn(attrstr, tmp, j);

    } else {
      sv_catpvn(attrstr, key, keylen);
    }

    /* Add '="value"' part if present*/
    if (SvOK(val)) {
      sv_catpvn(attrstr, "=\"", 2);
      GF_generate_attribute_value(attrstr, val);
      sv_catpvn(attrstr, "\"", 1);
    }
  }

  /* warn("real len: %d, %s", SvCUR(attrstr), SvPV_nolen(attrstr)); */

  return attrstr;
}

SV * GF_generate_tag(SV * tag, HV * attrhv, SV * val, int b_escapeval, int b_addnewline, int b_closetag) {
  char * tagsp, * valsp;
  STRLEN taglen, vallen, estimatedlen;
  SV * tagstr, * attrstr = 0;

  /* Force tag to string when getting length */
  tagsp = SvPV(tag, taglen);
  estimatedlen = taglen + 3 + (b_addnewline ? 1 : 0);

  /* Create attributes as string */
  if (attrhv) {
    attrstr = GF_generate_attributes(attrhv);
    estimatedlen += SvCUR(attrstr) + 1;
  }

  if (val) {
    /* If asked to escape, escape the val */
    if (b_escapeval)
      val = GF_escape_html(val, 0, 0, 0, 0);
    /* Force value to string when getting length */
    valsp = SvPV(val, vallen);
    estimatedlen += vallen + taglen + 3;
  }

  /* If asked to close the tag, add ' /' */
  if (b_closetag)
    estimatedlen += 2;

  /* Create new string to put final result in */
  tagstr = newSV(estimatedlen);
  SvPOK_on(tagstr);

  sv_catpvn(tagstr, "<", 1);
  sv_catsv(tagstr, tag);
  if (attrstr) {
    sv_catpvn(tagstr, " ", 1);
    sv_catsv(tagstr, attrstr);
    SvREFCNT_dec(attrstr);
  }
  if (b_closetag)
    sv_catpvn(tagstr, " />", 3);
  else 
    sv_catpvn(tagstr, ">", 1);

  if (val) {
    sv_catsv(tagstr, val);
    if (b_escapeval)
      SvREFCNT_dec(val);
    sv_catpvn(tagstr, "</", 2);
    sv_catsv(tagstr, tag);
    sv_catpvn(tagstr, ">", 1);
  }

  if (b_addnewline)
    sv_catpvn(tagstr, "\n", 1);

  return tagstr;
}

static char * hexlookup = "0123456789ABCDEF";

SV * GF_escape_uri(SV * str, SV * escchars, int b_inplace) {
  int i;
  STRLEN origlen, esclen, extrachars;
  char * sp, *newsp, *escsp;
  unsigned char c;
  SV * newstr;

  /* Get string pointer and length (in bytes) */
  if (b_inplace) {
    sp = SvPV_force(str, origlen);
  } else {
    sp = SvPV(str, origlen);
  }

  escsp = SvPV(escchars, esclen);

  /* Calculate extra space required */
  extrachars = 0;
  for (i = 0; i < origlen; i++) {
    c = (unsigned char)sp[i];
    /* Always escape control on 8-bit chars or chars in our escape set */
    if (c <= 0x20 || c >= 0x80 || memchr(escsp, c, esclen)) {
      extrachars += 2;
    }
  }

  /* Create new SV, or grow existing SV */
  if (b_inplace) {
    newstr = str;
    /* Always turn of utf8-ness in escaped string */
    SvUTF8_off(newstr);
    SvGROW(newstr, origlen + extrachars + 1);
  } else {
    newstr = newSV(origlen + extrachars + 1);
    SvPOK_on(newstr);
  }

  /* Set the length of the string */
  SvCUR_set(newstr, origlen + extrachars);

  /* Now do actual replacement (need to work
     backward for inplace change to work */

  /* Original string might have moved due to grow */
  sp = SvPV_nolen(str);

  /* Null terminate new string */
  newsp = SvPV_nolen(newstr) + origlen + extrachars;
  *newsp = '\0';

  for (i = origlen-1; i >= 0; i--) {
    c = (unsigned char)sp[i];
    if (c <= 0x20 || c >= 0x80 || memchr(escsp, c, esclen)) {
      newsp -= 3;
      newsp[0] = '%';
      newsp[1] = hexlookup[(c>>4) & 0x0f];
      newsp[2] = hexlookup[c & 0x0f];
    } else
      *--newsp = (char)c;
  }

  if (newsp != SvPV_nolen(newstr)) {
    croak("Unexpected length mismatch");
    return 0;
  }

  return newstr;
}

int GF_is_known_entity(char * sp, int i, int origlen, int *maxlen) {
  int start = i;

  if (++i < origlen) {
    /* Check for unicode ref (eg &#1234;) */
    if (sp[i] == '#') {
      int is_hex = 0;

      /* Check for hex unicode ref (eg &#x12af;) */
      if (i+1 < origlen && (sp[i+1] == 'x' || sp[i+1] == 'X')) {
        is_hex = 1;
        i++;
      }

      /* Not quite right, says "&#" and "&#;" are ok */
      while (++i < origlen) {
        if (sp[i] >= '0' && sp[i] <= '9') continue;
        if (is_hex && ((sp[i] >= 'a' && sp[i] <= 'f') || (sp[i] >= 'A' && sp[i] <= 'F'))) continue;
        if (sp[i] == ';' || sp[i] == ' ') {
          /* Keep track of maximum entity length */
          i++;
          if (maxlen && (i - start > *maxlen)) *maxlen = i-start;
          return 1;
        }
        break;
      }

    /* Check for entity ref (eg &nbsp;) */
    } else if ((sp[i] >= 'a' && sp[i] <= 'z') || (sp[i] >= 'A' && sp[i] <= 'Z')) {
      while (++i < origlen) {
        if ((sp[i] >= 'a' && sp[i] <= 'z') || (sp[i] >= 'A' && sp[i] <= 'Z')) continue;
        /* We should check to see if matched text string is known enity,
           but it's not that important */
        if (sp[i] == ';' || sp[i] == ' ') {
          /* Keep track of maximum entity length */
          i++;
          if (maxlen && (i - start > *maxlen)) *maxlen = i-start;
          return 1;
        }
        break;
      }
    }
  }
  return 0;
}

int GF_estimate_attribute_value_len(SV * val) {
  STRLEN vallen;
  I32 valtype;

  /* If reference, de-reference ... */
  if (SvROK(val)) {
    val = SvRV(val);
  }

  valtype = SvTYPE(val);

  /* Array case */
  if (valtype == SVt_PVAV) {
    int estimatedlen = 0;
    AV * aval = (AV *)val;
    I32 alen = av_len(aval), i;
    for (i = 0; i <= alen; i++) {
      SV **av_val;
      if ((av_val = av_fetch(aval, i, 0)) && SvOK(val = *av_val)) {
        estimatedlen += GF_estimate_attribute_value_len(val) + 1;
      }
    }
    return estimatedlen;
  }

  /* Hash case */
  if (valtype == SVt_PVHV) {
    int estimatedlen = 0;
    HV * hval = (HV *)val;
    char * key; I32 keylen;
    hv_iterinit(hval);
    while ((val = hv_iternextsv(hval, &key, &keylen))) {
      estimatedlen += keylen + 1;
    }
    return estimatedlen;
  }

  /* Ignore other non-scalar types */
  if (!SvOK(val)) return 0;

  /* Most common case of a string */
  if (SvPOK(val)) return SvCUR(val);

  /* Other SV case, turn it into a string */
  if (SvOK(val)) return (SvPV(val, vallen), vallen);

  return 0;
}

void GF_generate_attribute_value(SV * attrstr, SV * val) {
  I32 valtype;
  int no_escape = 0;

  /* If reference, de-reference ... */
  if (SvROK(val)) {
    val = SvRV(val);
    no_escape = 1;
  }

  valtype = SvTYPE(val);

  /* Array? Iterate over array items space separated... */
  if (valtype == SVt_PVAV) {
    AV * aval = (AV *)val;
    I32 alen = av_len(aval), i;
    for (i = 0; i <= alen; i++) {
      SV **av_val;
      if ((av_val = av_fetch(aval, i, 0)) && SvOK(val = *av_val)) {
        GF_generate_attribute_value(attrstr, val);
        if (i != alen) sv_catpvn(attrstr, " ", 1);
      }
    }
    return;
  }

  /* Hash? Iterate over keys space separated... */
  if (valtype == SVt_PVHV) {
    HV * hval = (HV *)val;
    char * key; I32 keylen;
    I32 hlen = hv_iterinit(hval), i = 0;
    HE * hentry;
    while ((hentry = hv_iternext(hval))) {
      key = hv_iterkey(hentry, &keylen);
      sv_catpvn(attrstr, key, keylen);
      if (++i != hlen) sv_catpvn(attrstr, " ", 1);
    }
    return;
  }

  /* Ignore other non-scalar types */
  if (!SvOK(val)) return;

  /* Otherwise just append to attribute string */

  /* If value was reference, use that unescaped */
  if (no_escape) {
    sv_catsv(attrstr, val);

  /* For the value part, escape special html chars, then dispose of result */
  } else {
    val = GF_escape_html(val, 0, 0, 0, 0);
    sv_catsv(attrstr, val);
    SvREFCNT_dec(val);
  }

  return;
}



lib/HTML/GenerateUtil.pm
package HTML::GenerateUtil;

use strict;
use warnings;

require Exporter;

our @ISA = qw(Exporter);

# This allows declaration	use HTML::GenerateUtil ':all';
# If you do not need this, moving things directly into @EXPORT or @EXPORT_OK
# will save memory.
our %EXPORT_TAGS = (
  'all' => [ qw(
    escape_html generate_attributes generate_tag escape_uri
    EH_INPLACE EH_LFTOBR EH_SPTONBSP EH_LEAVEKNOWN
    GT_ESCAPEVAL GT_ADDNEWLINE GT_CLOSETAG
    EU_INPLACE 
    $H a div span label ul ol li h1 h2 h3 h4
  ) ],
  'consts' => [ qw(
    EH_INPLACE EH_LFTOBR EH_SPTONBSP EH_LEAVEKNOWN
    GT_ESCAPEVAL GT_ADDNEWLINE GT_CLOSETAG
    EU_INPLACE
  ) ]
);

our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );

our @EXPORT = qw(
	
);

our $VERSION = '1.11';

our $H = 'HTML::GenerateUtil';

require XSLoader;
XSLoader::load('HTML::GenerateUtil', $VERSION);

use constant EH_INPLACE => 1;
use constant EH_LFTOBR => 2;
use constant EH_SPTONBSP => 4;
use constant EH_LEAVEKNOWN => 8;

use constant GT_ESCAPEVAL => 1;
use constant GT_ADDNEWLINE => 2;
use constant GT_CLOSETAG => 4;

use constant EU_INPLACE => 1;

# Preloaded methods go here.

my $escape_all = '"#$%&+,/:;<=>?@[]^`{}|\\' . "\x7f";
my $escape_lite = '"$+,/:;<=>@[]^`{}|\\' . "\x7f";

sub escape_uri { return escape_uri_internal($_[0], $_[2] || $escape_all, $_[1] || 0) }
sub escape_uri_lite { return escape_uri_internal($_[0], $_[2] || $escape_lite, $_[1] || 0) }

# If an unknown function is called, fill in some parameters and
# call generate_tag
# e.g. 
#                 font( $html )  
#      maps to    generate_tag('font',undef,$html, 0 )
#                 font( { size => 1 }, $html )  
#      maps to    generate_tag('font',{ size => 1},$html, 0 )
#                 font( $html, GT_ADDNEWLINE )  
#      maps to    generate_tag('font',{ size => 1},$html, GT_ADDNEWLINE )
sub AUTOLOAD {

  # assume the function name is the tag name
  my $Tag = our $AUTOLOAD;
  $Tag =~ s{.*::}{};

  # if the function was called on the class name, strip out the class name
  shift if $_[0] eq $H;

  # if the first parameter was not a ref, assume no attributes passed,
  # so we use an empty attr list 
  unshift @_, undef unless ($_[0] && ref($_[0]));

  # Use the tag as the first parameter
  unshift @_, lc $Tag;

  # If no flags were specified, set to 0
  $_[3] ||= 0;

  goto &generate_tag;
}

1;

__END__

=head1 NAME

HTML::GenerateUtil - Routines useful when generating HTML output

=head1 SYNOPSIS

  use HTML::GenerateUtil qw(escape_html generate_attributes generate_tag escape_uri :consts $H div);

  my $Html = "text < with > things & that need \x{1234} escaping";
  $Html = escape_html($Html);

  ... or ...

  escape_html($Html, EH_INPLACE);

  ... also ...

  my $Attr = generate_attributes({ href => 'http://...', title => 'blah' });
  $Html = "<a $Attr>$Html</a>";

  ... but even better ...

  $Html = generate_tag('a', { href => 'http://...', title => 'blah' }, $Html, 0);

  ... also you might want something like ...

  my $URI = 'http://host/?' . join ";", map { $_ => escape_uri($Params{$_}) } keys %Params;
  $Html = generate_tag('a', { href => $URI }, $Html, 0);

  ... you can shortcut that by importing a function, or using the autoloading $H object ...

  div({ class => [ qw(a b) ] }, "div content");
  $H->a({ href => $URI  }, "text", GT_ADDNEWLINE);

=head1 DESCRIPTION

Provides a number of functions that make generating HTML output
easier and faster. All written in XS for speed.

=head1 CONTEXT

When creating a web application in perl, you've got a couple of main choices on how
to actually generate the HTML that gets output:

=over 4

=item *

Programatically generating the HTML in perl

=item *

Using some template system for the HTML and inserting the data
calculated in perl as appropriate

=back

Your actual application, experience and environment will generally determine
which is the best way to.

If you go the programatic route, then you generally need some way of
generating the actual HTML output in perl. Again, there's generally a couple
of ways of doing this.

=over 4

=item *

Just joining together text strings in perl as appropriate.

  Eg. $link = "<a href="$ref">$text</a>";

=item *

Or using some function module like CGI

  Eg. $line = a({ href => $ref }, $text);

=item *

More complex object systems like HTML::Table

=back

The first seems easy, but it gets harder when you have to manually escape
each string to avoid placing special HTML chars (eg E<lt>, etc) in strings
like $text above.

With the CGI, most of this is automatically taken care of, and most
strings are automatically escaped to replace special HTML chars with
their entity equivalents.

While this is nice, CGI is written in pure perl, and can end up being a
bit slow, especially if you already have a fast system that generates
pages very heavy in tags (eg lots of table elements, links, etc)

That's where this module comes it. It provides functions useful for
escaping html and generating HTML tags, but it's all written in XS to
be very fast. It's also fully UTF-8 aware.

=head1 FUNCTIONS

=over 4

=item C<escape_html($Str [, $Mode ])>

Escapes the contents of C<$Str> to change the chars
[<>&"] to '&lt;', '&gt;', '&amp;' and '&quot;' repectively.

C<$Mode> is an optional bit field with the additional options or'd together:

=over 4

=item *

C<EH_INPLACE> - modify in-place, otherwise return new copy

=item *

C<EH_LFTOBR> - convert \n to <br>

=item *

C<EH_SPTONBSP> - convert '  ' to ' &nbsp;'

=item *

C<EH_LEAVEKNOWN> - if & is followed by text that looks like an
entity reference (eg &#1234; or &#x1ab2; or &nbsp;) then it's
left unescaped

=back

Useful for turning text into similar to <pre> form without
actually being in <pre> tags

=item C<generate_attributes($HashRef)>

Turns the contents of C<$HashRef> of the form:

  {
    aaa => 'bbb',
    ccc => undef
  }

Into a string of the form:

  q{aaa="bbb" ccc}

Useful for generating HTML tags. The I<values> of each hash
entry are escaped with escape_html() before being added
to the final string.

If you want to use a raw value unescaped, pass it as a
scalar ref with a single item. Eg.

  {
    aaa => \'<blah>',
    bbb => '<blah>'
  }

Is turned into:

  q{aaa="<blah>" bbb="&lt;blah&gt;"}

If the value is an array ref, then the individual items
are joined together with a space separator. Eg.

  {
    class => [ 'class1', 'class2', \'<blah>' ],
    aaa => 'bbb'
  }

Is turned into:

  q{aaa="bbb" class="class1 class2 <blah>"}

If the value is a hash ref, then the individual keys
are joined together with a space separator. Eg.

  {
    class => { class1 => 1, class2 => 2 ],
    aaa => 'bbb'
  }

Is turned into:

  q{aaa="bbb" class="class2 class1"}

Keys are always escaped since you can't have a scalar
reference as a key.

=item C<generate_tag($Tag, $AttrHashRef, $Value, $Mode)>

Creates an HTML tag of the basic form:

  <$Tag %$AttrHashRef>$Value</$Tag>

If C<$AttrHashRef> is C<undef>, then no attributes are created.
Otherwise C<generate_attributes()> is called to stringify
the hash ref.

If C<$Value> is C<undef>, then no C<$Value> is included, and
no E<lt>/$TagE<gt> is added.

C<$Mode> is a bit field with the additional options:

=over 4

=item *

C<GT_ESCAPEVAL> - call escape_html on $Value

=item *

C<GT_ADDNEWLINE> - append \n to output of string

=item *

C<GT_CLOSETAG> - close the tag (eg <tag />). This should really
only be used when C<$Value> is undef, otherwise you'll end
up with something like C<E<lt>tag /E<gt>valueE<lt>/tagE<gt>>,
which is probably not what you want

=back

=item C<escape_uri($Uri, [ $Mode, $EscapeChars ])>

Escape unsafe characters in a uri.

This escapes all characters not in the unreserved character set.
As a regexp that is:

  [^A-Za-z0-9\-_.!~*'()]

or

  [\x00-\x1F "#$%&+,/:;<=>?@\[\]^`{}|\\\x7f-\xff];

Some other things to note:

=over 4

=item *

The escaping assumes all strings with high-chars are utf-8 strings. That
is it first turns off any utf-8 bit on the string, and then encodes each
byte to it's corresponding octet.

=item *

When encoding a uri with parameters, you'll probably want to encode each
parameter first and then join it to the final string, something like:

  my %uri_params = ( ... )
  my $uri = "http://hostname.com/somepath/?" .
    join ";",
    map { $_ . "=" . escape_uri($uri_params{$_}) } 
    keys %uri_params;

Assuming your keys don't have any unreserved characters in them, a common
practice in many peoples design.

Doing something like:

  my $uri = escape_uri("http://hostname.com/somepath/?a=p1");

Will escape the '?', not giving you what you expect at the other end.

=back

C<$Mode> is a bit field with the additional options:

=over 4

=item *

C<EU_INPLACE> - modify in-place, otherwise return new copy

=back

=back

=cut

=head1 BUGS AND LIMITATIONS

The EH_LEAVEKNOWN option is just heuristic, and accepts anything
that even looks like an entity reference, even if it isn't a
correct one. I'm not sure if this is a securit issue or not.

=cut

=head1 SEE ALSO

L<Apache::Util>, L<HTML::Entities>, L<CGI>

Latest news/details can also be found at:

L<http://cpan.robm.fastmail.fm/htmlgenerateutil/>

Available on github at:

L<https://github.com/robmueller/html-generateutil/>

=cut

=head1 AUTHOR

Rob Mueller E<lt>cpan@robm.fastmail.fmE<gt>

=cut

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2004-2011 by Opera Software Australia Pty Ltd

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself. 

=cut

